common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -208,9 +228,16 @@ public void invoke(@NonNull final Result<UsbSmartCardConnection, IOException> va
                     final PivSession piv = new PivSession(c);
                     //Check if there are no PIN attempts remaining.
                     if (piv.getPinAttempts() == 0) {
-                        showMaxAttemptsDialog();
+                        showErrorDialog(R.string.smartcard_max_attempt_dialog_title, R.string.smartcard_max_attempt_dialog_message);
```

Comment:
Not related to this line

I'm wondering if we need to call requestConnection twice (on line 222 and 423). Can we reuse the  PivSession object (since this whole ChallengeHandler is ONE session) on line 222 in 423's operation?

If we can do this, then we'll be able to get rid of mDevice null checks and duplicated connection codes.

Reply:
Even if that's not the case... we probably could still refactor the connection code and its error handler to be a separate method like this.

```Java

public interface IPivSessionCallback {
    void onGetSession(@NonNull final PivSession p);
}


private void getActivePivSessionAsync(@NonNull final ClientCertRequest request, @NonNull final IPivSessionCallback callback){
    if (mDevice == null) {
        showErrorDialog();
        request.cancel();
        return;
    }

    mDevice.requestConnection(UsbSmartCardConnection.class, new Callback<Result<UsbSmartCardConnection, IOException>>() {
        @Override
        public void invoke(@NonNull final Result<UsbSmartCardConnection, IOException> value) {
            try {
                final SmartCardConnection c = value.getValue();
                final PivSession piv = new PivSession(c);

                // validate is piv session is valid, i.e. check pin attempt, etc. 

                callback.onGetSession(piv);
            } catch (.... Exceptions){
                showErrorDialog();
                request.cancel();
            }
        }
    });
}
```

Usage
``` Java
private void handleSmartcardCertAuth(@NonNull final ClientCertRequest request) {
    getActivePivSessionAsync(request, new IPivSessionCallback() {
        @Override
        void onGetSession(@NonNull final PivSession p){
            //load certs.
            //display certs.
        }
    });
}
```

Reply:
getActivePivSessionAsync() should also be reusable in getPivProviderCallback();

Reply:
> Not related to this line
> 
> I'm wondering if we need to call requestConnection twice (on line 222 and 423). Can we reuse the PivSession object (since this whole ChallengeHandler is ONE session) on line 222 in 423's operation?
> 
> If we can do this, then we'll be able to get rid of mDevice null checks and duplicated connection codes.

I'm not totally sure if ChallengeHandler is entirely in one session. 
My experience with this before was that I attempted to create an instance variable to hold the PivSession, but when I attempted to use it outside of the invoke method via a button click callback (which I believe all of the dialog button callbacks are outside of the initial invoke method), I got the `Failed to send 25 bytes` error we've seen before. I don't 100% remember if I tried passing the PivSession object as a parameter, but I wouldn't be surprised if I already tried it (I recall spending a long time trying to debug this).

I will do the `IPivSessionCallback` refactoring though.